packages/auth/src/session.ts

```diff
@@ -0,0 +1,302 @@
+    this.jwts = jwts ? jwts : {}
+  }
+
+  get name(): string {
+    return this.metadata.name
+  }
+
+  get expiration(): number {
+    return this.metadata.expiration ? Math.max(this.metadata.expiration, 120) : 3e7
+  }
+
+  onAuth(cb: () => void) {
+    this.onAuthCallbacks.push(cb)
+  }
+
+  setAccount(account: Account) {
+    this.account = account
+  }
+
+  setConfig(config: WalletConfig) {
+    this.config = config
+  }
+
+  static async open(args: {
+    context: WalletContext,
+    networks: NetworkConfig[],
+    referenceSigner: string,
+    signers: { signer: Signer, weight: ethers.BigNumberish }[],
+    thershold: ethers.BigNumberish,
+    metadata: SessionMeta,
+    deepSearch?: boolean,
+    knownConfigs?: WalletConfig[],
+    noIndex?: boolean
+  }): Promise<Session> {
+    const {
+      context,
+      networks,
+      referenceSigner,
+      signers,
+      thershold,
+      deepSearch,
+      knownConfigs,
+      noIndex,
+      metadata
+    } = args
+
+    const authChain = networks.find((n) => n.isAuthChain)
+    if (!authChain) throw Error('Auth chain not found')
+  
+    const authProvider = authChain.provider ? authChain.provider : new ethers.providers.JsonRpcProvider(authChain.rpcUrl)
+    const configFinder = new SequenceUtilsFinder(authProvider)
+  
+    const solvedSigners = Promise.all(signers.map(async (s) => ({ ...s, address: await s.signer.getAddress() })))
+  
+    const existingWallet = (await configFinder.findLastWalletOfInitialSigner({
+      signer: referenceSigner,
+      context: context,
+      provider: authProvider,
+      requireIndex: deepSearch ? false : true
+    })).wallet
+
+    if (existingWallet) {
+      // Find prev configuration
+      const config = (await configFinder.findCurrentConfig({
+        address: existingWallet,
+        provider: authProvider,
+        context: context,
+        knownConfigs
+      })).config
+  
+      if (!config) throw Error('Wallet config not found')
+  
+      // Load prev account
+      const account = new Account({
+        initialConfig: config,
+        networks: networks,
+        context: context
+      }, ...signers.map((s) => s.signer))
+  
+      const session = new Session(
+        config,
+        context,
+        account,
+        metadata 
+      )
+
+      // Fire JWT requests before opening session
+      networks.map((n) => session.scheduleAuth(n))
+
+      const [newConfig] = await account.updateConfig(
+        editConfig(config, {
+          threshold: thershold,
+          set: await solvedSigners
+        }), noIndex ? false : true
+      )
+
+      session.setConfig(newConfig)
+      session.setAccount(new Account({
+        initialConfig: newConfig,
+        networks: networks,
+        context: context
+      }, ...signers.map((s) => s.signer)))
+
+      // Fire JWT requests again, but with new config
+      // MAYBE This is not neccesary, we can rely on the first request?
+      networks.forEach((n) => session.scheduleAuth(n))
```

Comment:
yes we can rely on the first one. I think for ethauth to get a jwt token -- we can just submit a partial signing from the login key or any key in the wallet config prior to this new session. Is it possible with for us to do partial signature validation with 1271? the idea is ethauth is used on both client and server as a signing of eip712 payload and then validation on server. On ethauth.js (v0.4.3), we don't need to validate the signature upon signing, we can just do a partial signing of the eip712 and fire it off to the server for exchange for a jwt. We have go-ethauth on server side which can also sign or validate a eip712 "claim" / "proof", and here we could validate a partial signature and in the response of the proof we could return the signers which have successfully validated the 712 ethauth claim. Then on the jwt handler where we take an ethauth proof and exchange for a jwt, we can check the signer list and see if the threshold is at least up to one gold key, at which point, we will issue the jwt.

Reply:
I have a better solution to removing this secondary auth schedule here: https://github.com/0xsequence/sequence.js/pull/102/commits/a319d3795177de55e1c2cf5e30e814bd6e7571e2#diff-da2446c8cacd86376976dc5132ed9d5de20688e372f61f80d15f02f3061261fbR129-R143

+ makes a number of other improvements and performance updates as a result

Reply:
The issue is not that we are over-signing with signers that later become invalid, the issue is that Sequence signature encoding and wallet configuration is intertwined, so when you update the configuration of the wallet, old signatures that should still be valid become invalid because the encoding changes, for example:

```
Original configuration:
Signer A - Signer B - Signer C

Signed Transaction 1:
Signature A - Signature B - Address C

Change configuration to:
Signer A - Signer B - Signer D

Now the signed transaction 1 is no longer valid, because it should be encoded as:
Signature A - Signature B - Address D
```

There are two man issues to solve it:

1 - EIP1271 signatures don't have any way of validating "partial signatures".
2 - The JWT service doesn't have a way of retrieving the contents of the latest configuration of the wallet, it only has the "hash" of the configuration and thus needs the full configuration in some way to validate a signature.

This being a race condition I think we can find more simple solutions, one could be waiting for the JWT token before updating the wallet (this is what happens most of the time anyway, the second "try" is more to catch a rare race condition when the wallet updates faster than the JWT generation), this should work 100% of the time and it's clean, but it would increase the time it takes to open the session.

Another option is to send a specific "block" that should be used to check the signature, then the JWT server can validate the signature using the wallet at the state of the given block, effectively also enforcing order but without blocking. There are two details we need to keep in mind before going this route:

1) The JWT server would need to have some limit on how old the requested block can be, otherwise it would open a secuirty issue where an attacker can get a JWT token using an old configuration.
2) The JWT client would need to get the current block of the network before sending the JWT request, the server would also need to know the current block of the network to check the age of the block, those two things combined could make this process more slow than just waiting for the original JWT as proposed before.